#1675  update inProgress Call CDRs to completed status

anhntnguyen opened the pull request:
Hi @maria-farooq ,

Could you please help me to review this pull request. I have tested it manually and the results are good, all call status like Ringing, in-progress, queued completely changes status to COMPLETED, the rest of status is un-changed.

maria-farooq commented:
related to #1675

maria-farooq commented on restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/CallManager.java:
```diff
@@ -281,6 +283,27 @@ public CallManager(final Configuration configuration, final ServletContext conte
+        Sid sid = new Sid(instanceId);
+        final CallDetailRecordsDao callDetailRecordsDao = storage.getCallDetailRecordsDao();
+        List<CallDetailRecord> results= callDetailRecordsDao.getInCompleteCallDetailRecordsByInstanceId(sid);
+        for (CallDetailRecord result : results) {
```
is there a way we can avoid this loop and lower the cost of this operation?

anhntnguyen replied:
I have reduce the cost by the SQL statement. I don't have any idea on how to avoid this loop. Do you have any idea ?

anhntnguyen replied:
@maria-farooq 
a ha, I got your point now. in fact, instead of trying to query a database back to the node as I'm using now, I need to update it directly.
Thanks a lot Maria :)

anhntnguyen replied:
We are having UserAgentmanager class, would be the same as this one and it use intergradtion test for it. my new function  firstTimeCleanup is void function. I still dont know how to create unit test for it. It would be nice if you can give my some idea on how to create unit test here.

maria-farooq commented on restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/CallManager.java:
```diff
@@ -281,6 +283,27 @@ public CallManager(final Configuration configuration, final ServletContext conte
+        for (CallDetailRecord result : results) {
+            CallDetailRecord updatedResult = result.setStatus(CallStateChanged.State.COMPLETED.name());
+            if (logger.isDebugEnabled()) {
+                    logger.debug("Call From: " + updatedResult.getFrom() + " change status from " + result.getStatus() + " to COMPLETED");
```
Would it make sense to add callSid in this log statement?

anhntnguyen replied:
Agree, will update it to callSid

maria-farooq commented:
Hi @anhntnguyen do we have any update here

anhntnguyen commented:
Hi @maria-farooq 

I just updated the code, but also waiting for your response on unit test  that I have asked before.
I have updated the sql statement to update all in-complete calls on SQL side.
For unittest: 
this is void funtion, so that I could not test it via unit test. It need to be tested as we have on UserAgentmanager  class. Or do you have any idea , please suggest to me:)

Best regards
Anh Nguyen

anhntnguyen commented:
@maria-farooq 
I have just updated the xml file, please help me to review :)

maria-farooq commented on restcomm/restcomm.application/src/main/webapp/WEB-INF/scripts/mariadb/sql/call-detail-records.xml:
```diff
@@ -181,6 +181,11 @@
 	</select>
 
+  <select id="getInCompleteCallDetailRecordsByInstanceId" parameterType="string" resultType="hashmap">
+    SELECT * FROM "restcomm_call_detail_records" WHERE "instanceid"=#{instanceid} AND (UPPER("status") = ('IN_PROGRESS') OR 
```
@anhntnguyen 
did we tested this file?
I think double quotes are not supported in this mariadb sql formats in this file.
i mean double quotes around table and column names.

maria-farooq commented on restcomm/restcomm.application/src/main/webapp/WEB-INF/scripts/mariadb/sql/call-detail-records.xml:
```diff
@@ -201,4 +206,11 @@
   </update>
+
+  <update id="updateInCompleteCallDetailRecordsToCompletedByInstanceId" parameterType="string">
+    UPDATE "restcomm_call_detail_records"
```
@anhntnguyen 
did we tested this file?
I think double quotes are not supported in this mariadb sql formats in this file.
i mean double quotes around table and column names.

maria-farooq commented:
Hi @anhntnguyen thanks for your changes.
Did you tested new xml file local system?
I think double quotes are not supported in this file.
I mean double quotes around table and column names.
Thanks
Maria

maria-farooq commented:
Hello @anhntnguyen ,

For testing mariadb sql file, we can setup MySql and integrate it with Restcomm.
Please see the documentation [here](http://documentation.telestax.com/connect/configuration/How%20to%20get%20started%20with%20Restcomm-Connect%20and%20Mysql.html).
Please feel free to ask if any questions

Thanks
Maria

anhntnguyen commented:
Hi @maria-farooq ,

I'm really sorry for delay it quite long. I'm just back from new year vacation. I have tested on mariadb with mysql and " is not accepted in that case. Thank you for giving me a hint on that. I have changed the sql statement and tested it with mysql, everything is working well now. Please let me know if you find anything could be a bug, I will try to fix it :)

Best Regards
Anh Nguyen

maria-farooq commented:
thanks @anhntnguyen for you changes.
this branch has conflicts with master, can you please resolve them so we can move forward on merge.
Thanks
Maria

anhntnguyen commented:
Hi @maria-farooq,

The conflict is gone now :)

anhntnguyen commented:
Hi @maria-farooq 

I have done for conflict resolving :)

maria-farooq commented:
thanks @anhntnguyen i will review it

maria-farooq commented:
Hi @anhntnguyen my sincerest apologies for the delay in reviewing PR.
I will review it next week.
thanks
Maria

maria-farooq commented:
Hi @gvagenas ,

I have reviewed this PR and seems fine to me.

Would you like to give a final view please.

Thanks 
Maria

gvagenas commented:
Hi @anhntnguyen , thanks for your contribution.
@maria-farooq what is the level of importance for this issue? Is it Major or Critical level?

maria-farooq commented:
@gvagenas We have an issue which causes CallManager actor to restart over exception.

I am afraid we can not merge this PR before we solve the issue and prevent CallManager to restart on exception in `onReceive`

Otherwise this patch can cause moving call that are actually `in-progress` to `completed` state.

references:https://github.com/RestComm/Restcomm-Connect/issues/1897
